proposals/storage-get-keys.md

```diff
@@ -0,0 +1,115 @@
+# Proposal: browser.storage.&lt;area>.getKeys
+
+**Summary**
+
+API to retrieve all keys for a given storage area in the `browser.storage` API.
+
+**Document Metadata**
+
+**Author:** [Oliver Dunk (Google)](https://github.com/oliverdunk)
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** [polywock](https://github.com/polywock)
+
+**Created:** 2024-06-28
+
+**Related Issues:** [#601](https://github.com/w3c/webextensions/issues/601)
+
+## Motivation
+
+### Objective
+
+This API enables developers to retrieve the keys in storage without needing to
+receive all of the data stored alongside those keys. Depending on the
+implementation, the impact of this optimization may differ - for example
+Chromium has only a limited ability to access data from storage without
+fetching the associated data. Regardless, this reduces the amount of data
+returned to the caller and allows for further optimizations in the future.
+
+#### Use Cases
+
+There are existing use cases for enumerating the data in extension storage. For
+example, extensions for Chrome that add extension storage areas to DevTools.
+While these also need a way to show the associated values, they may not want to
+load these upfront.
+
+Additionally, in some cases developers may want to access keys with a given
+prefix, for example if using a key like `foo:tabid`. This API proposal does not
+allow matching by prefix (see "Future Work") but developers would be able to do
+this themselves after obtaining the list of keys.
```

Comment:
Maybe it's gonna be too specific but would you consider adding a parameter for prefix(es)? Null or undefined or a string or an array of strings, just like browser.storage.&lt;area>.get and continue to return an array of strings (or an empty array).

I'm in this group (who wants to access keys with a given prefix, actually prefixes). I didn't know this feature was added. I was using browser.storage.&lt;area>.getBytesInUse and to give that as parameter I was calling the whole storage and extracting only the keys I needed by calling Object.keys.filter. When the storage was taking up space (~25MB) I realized this process was sloving down the page and I googled for a more practical way. Then I found this feature in [here](https://stackoverflow.com/a/79179793/13630257).

I have a browser extension designed to be used on a few specific websites. The part that interests this feature is like this. The websites don't offer GIF avatars because of both server and client side load. But they allow uploads. The server takes up the first frame and resize it as JPEG. And the original files are accessible to everyone in specific URL format. The extension fetches the GIFS and resize them to provide smaller, non-burdeningclient-side. I keep those avatars as [siteId]_[userId] in the storage.

Even though they are small, they accumulate and take up space because I keep them in Base64. Because of this, I decided to put a size display and a reset option. Firefox doesn't currently offer browser.storage.&lt;area>.getBytesInUse and requires manual (slower) calculation. Although I don't use very large storage except for avatars, it would be much more convenient to give the prefix(es) directly and get the keys that way.

Reply:
Thanks for sharing your use case! As mentioned in this proposal, we're definitely open to some way of filtering the keys. I think there's still an open question about what that would look like. In theory, prefixes would allow for optimization and be the simplest approach. In practice, however, I doubt any of the existing implementations could query by prefix - they would instead be retrieving all of the keys and filtering. At that point, there's some (but less) benefit to using a prefix and more advanced matching might make sense.

I think opening a new issue here would be reasonable if you're interested. The hardest thing will likely be finding a sponsoring browser - I suspect we won't have the capacity to prioritize this on the Chrome side in the short term. We would always accept patches once we've agreed on an API though.